### ISSUE TITLE
Optional parameters with defaults can not be overridden by inputs.json

### ISSUE THREAD

**Issue by rhpvorderman** [broadinstitute/cromwell#3528]
Say I have a task called `bwaMem` in a workflow called `alignment`.

The task bwaMem looks as follows.
```WDL
task bwaMem {
  File in1
  File? in2
  String indexBase
  Int? threads = 1
  command {
   bwa mem ${"-t" +threads}  ${indexBase} ${in1} ${in2}
  }
  output {
    File samFile = glob(*.sam)
  }
 runtime {
  cpu = select_first([threads])  
  }
}
```
Then I can use the following inputs.json for the workflow `alignment`:
```JSON
{
  "alignment.bwaMem.threads": 3
}
```
But this is ignored. The default value of 1 is used.
On the contrary, if I set defaults for the `alignment` workflow itself. I can override these in the inputs.json. But apparently this does not work for imported tasks?

**Comment by Redmar-van-den-Berg** [comment]
Your workflow actually says that `threads` is an `optional int` that equals `1`. The notation you used does not mean that 1 is the default value for `threads` The correct way to give a default value for an optional parameter is as follows:
```
task bwaMem {
  File in1
  File? in2
  String indexBase
  Int? threads
  command {
   bwa mem -t ${default="1" threads} ${indexBase} ${in1} ${in2}
  }
  output {
    File samFile = glob(*.sam)
  }
 runtime {
  cpu = select_first([threads])  
  }
}
```

**Comment by rhpvorderman** [comment]
> Your workflow actually says that threads is an optional int that equals 1. The notation you used does not mean that 1 is the default value for threads

No this is actually the canonical way of defining defaults. It is explicit, and it makes sure the same default for threads is used in both the runtime and the command section.

There is actually another bug related to this issue:
https://github.com/broadinstitute/cromwell/issues/2532

**Comment by cjllanwarne** [comment]
I agree that optional task inputs not being overridable from the inputs json is a bug and should be fixed.

Regarding defaults vs fixing values, you're both right - values like that might be intended as inputs or might be intended as fixed intermediate values depending on the context and it's really hard to deduce which one an author intended just based on the WDL file. 

That kind of confusion is exactly why WDL 1.0 (you're currently writing in draft-2) is adding `input` sections. Ie:

```wdl
workflow foo {
  input {
    # input with default
    Int threads = 1
  }
}
```
vs
```wdl
workflow foo {
    # intermediate value: cannot be overridden
    Int threads = 1
}
```

**Comment by rhpvorderman** [comment]
> That kind of confusion is exactly why WDL 1.0 (you're currently writing in draft-2) is adding input sections. Ie:

Wow! That is some great functionality.I love how WDL is continuously identifying real problems, and finding solutions for them in a readable, logical and easily to understand syntax. :+1:

Will this inputs section solve this bug? Presumably it will make it at least easier to solve it for the Cromwell developers?

**Comment by rhpvorderman** [comment]
Closed as not an issue anymore in 1.0
